### ISSUE TITLE
Failing tests?

### ISSUE THREAD

**Issue by mindplay-dk** [InteractiveAdvertisingBureau/iabtcf-es#221]
**Version**

I've tried `master` as well as `tags/1.1.0`, I see failing tests on both.

**Describe with reproduction steps – What is the expected behavior?**

I run the following commands:

```
git clone https://github.com/InteractiveAdvertisingBureau/iabtcf-es.git
cd iabtcf-es
yarn
yarn build
yarn test
```

I see the following output:

```
yarn run v1.22.5
$ yarn && mocha `find ./modules/**/test/ '*.test.ts'`
[1/4] Resolving packages...
success Already up-to-date.
find: ‘*.test.ts’: No such file or directory

<---- snip ---->

  322 passing (1s)
  3 failing

  1) command->PingCommand
       should return a Ping object (and nothing else) when called:

      ping instanceof Ping
      + expected - actual

      -false
      +true

      at PingCommand.pingCallback [as callback] (modules/cmpapi/test/command/PingCommand.test.ts:11:65)
      at PingCommand.Command.invokeCallback (modules/cmpapi/src/command/Command.ts:51:12)
      at PingCommand.Command (modules/cmpapi/src/command/Command.ts:28:12)
      at new PingCommand (modules/cmpapi/src/command/PingCommand.ts:21:42)
      at Context.<anonymous> (modules/cmpapi/test/command/PingCommand.test.ts:17:5)
      at processImmediate (internal/timers.js:456:21)

  2) CmpApi
       should override the stub:
     TypeError: Cannot read property 'cmpId' of null
      at PingCommand.callback (modules/cmpapi/test/CmpApi.test.ts:100:21)
      at PingCommand.Command.invokeCallback (modules/cmpapi/src/command/Command.ts:51:12)
      at PingCommand.Command (modules/cmpapi/src/command/Command.ts:28:12)
      at new PingCommand (modules/cmpapi/src/command/PingCommand.ts:21:42)
      at CallResponder.apiCall (modules/cmpapi/src/CallResponder.ts:109:9)
      at /home/mindplay/workspace/iabtcf-es/modules/cmpapi/test/CmpApi.test.ts:98:22
      at new Promise (<anonymous>)
      at /home/mindplay/workspace/iabtcf-es/modules/cmpapi/test/CmpApi.test.ts:96:12
      at step (modules/cmpapi/test/CmpApi.test.ts:33:23)
      at Object.next (modules/cmpapi/test/CmpApi.test.ts:14:53)
      at fulfilled (modules/cmpapi/test/CmpApi.test.ts:5:58)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)

  3) CmpApi
       should call a built-in ping command through the page interface and decorate the response in the middleware:
     TypeError: Cannot create property 'u' on boolean 'false'
      at PingCommand._a.<computed> [as callback] (modules/cmpapi/test/CmpApi.test.ts:401:36)
      at PingCommand.Command.invokeCallback (modules/cmpapi/src/command/Command.ts:51:12)
      at PingCommand.Command (modules/cmpapi/src/command/Command.ts:28:12)
      at new PingCommand (modules/cmpapi/src/command/PingCommand.ts:21:42)
      at CallResponder.apiCall (modules/cmpapi/src/CallResponder.ts:105:9)
      at Context.<anonymous> (modules/cmpapi/test/CmpApi.test.ts:409:24)
      at processImmediate (internal/timers.js:456:21)



error Command failed with exit code 3.
```

Is `master` supposed to be stable? (a tagged release at least should have working tests, right?)

It took me a while to get this project to install and build.

I had to install `yarn` to get this to install and build. I use `npm` daily, and it doesn't look like the project works with `npm`? It would be really nice if the README contained some build instructions, and a list of global dependencies, such as `yarn`. It also wasn't clear to me that the `build` command had to be run before the `test` command - this should be in the README, too.

I'm on WSL 2 running Ubuntu 20.04.

**Comment by mindplay-dk** [comment]
I also tried running your CI configuration on Travis:

https://travis-ci.org/github/mindplay-dk/iabtcf-es/jobs/725578987

Here the tests are failing to start:

```
$ yarn run test:lcov
yarn run v1.22.4
$ nyc --reporter=lcov mocha
Error: No test files found: "test"
error Command failed with exit code 1.
```

Do you have a CI set up somewhere else? (it would be nice to have this information in the README as well.)

Where are dependencies such as `terser` supposed to come from? I see it in `yarn.lock`, but the only dependencies listed in your `package.json` are `json` and `mocha` - this is probably why it doesn't work without `yarn`? (Seems like something is missing here? Lock files aren't really an alternative to `devDependencies`, since they can't specify version constraints...)

**Comment by chrispaterson** [comment]
Hmm, that's strange because I have no such failure.  I followed your steps with a clean/new clone and got:

```
yarn run v1.22.4
$ yarn && mocha `find ./modules/**/test/ '*.test.ts'`
[1/4] 🔍  Resolving packages...
success Already up-to-date.
find: *.test.ts: No such file or directory


  response->TCData
    ✓ should create a TCData based on the TCModel with unrestricted vendors (65ms)
    ✓ should create a TCData based on the TCModel with vendors
    ✓ should create a TCData with only gdprApplies, tcfPolicyVersion, cmpId and cmpVersion if gdprApplies=false
    ✓ should encode purpose restrictions (63ms)

  response->InAppTCData
    ✓ should create a InAppTCData based on the TCModel with unrestricted vendors (40ms)
    ✓ should create a InAppTCData based on the TCModel with vendors
    ✓ should encode purpose restrictions (126ms)

  response->Response
    ✓ populates a response with CmpApiModel values

  response->Ping
    ✓ populates a response with CmpApiModel values
    ✓ sets gvlVersion if a TCModel has a vendorListVersion

  response->Disabled
    ✓ populates a response with CmpApiModel values

  CmpApiModel
    ✓ has these default values

  status->EventStatus
    ✓ should have these values

  status->CmpStatus
    ✓ should have these values

  status->DisplayStatus
    ✓ should have these values

  Reported issues
    ✓ Issue 96 CmpApi should respond to addEventListener call with an error object when in an Error state (47ms)
    ✓ Should not throw an error if getTCData is called before update is called (41ms)
    ✓ 126 Should not throw an error tc string is set as ""
    ✓ 120 AddEventListener: 'cmpuishown' isn't being triggered - test toggles update UIVisible (77ms)
    ✓ 164 Queued 'getTCData' failing when vendorIds is set
    ✓ 195 getInAppTCData no longer returns a success argument
    ✓ 196 should not call callback twice on addEventListener
    ✓ 208 Custom Commands no longer get called before update

  EventListenerQueue
    ✓ should add and remove a listener when add() and remove() are called respectively
    ✓ should clear all listeners when clear() is called

  CmpApi
    ✓ should override the stub
    ✓ should throw an error if cmpId=0 and cmpVersion=2
    ✓ should throw an error if cmpId=1 and cmpVersion=2
    ✓ should throw an error if cmpId=null and cmpVersion=2
    ✓ should throw an error if cmpId=banana and cmpVersion=2
    ✓ should throw an error if cmpId=0.234 and cmpVersion=2
    ✓ should throw an error if cmpId=-1 and cmpVersion=2
    ✓ should throw an error if cmpId=2 and cmpVersion=2
    ✓ should throw an error if cmpId=2.0 and cmpVersion=2
    ✓ should throw an error if cmpId=2 and cmpVersion=-1
    ✓ should throw an error if cmpId=2 and cmpVersion=2.02345
    ✓ should throw an error if cmpId=2 and cmpVersion=0
    ✓ should throw an error if cmpId=2 and cmpVersion=banana
    ✓ should throw an error if cmpId=2 and cmpVersion=null
    ✓ should pick up a queued stub function an excecute it if it can when tcModel is set (47ms)
    ✓ should set cmpStatus to "error" if disabled is set to true
    ✓ should pick up a queued stub function an excecute it if it can when tcString is set
    ✓ should set CmpApiModel.displayStatus = "visible" when uiVisible is passed as true to the update function
    ✓ should set CmpApiModel.dsiabled to true when CmpApi.disable() is called
    ✓ should throw an error when disabled and update is called
    ✓ should callback with success=false and result=error message if command=foo and version=2 because it's and unknown command
    ✓ should callback with success=false and result=error message if command=5 and version=2 because command is a number
    ✓ should callback with success=false and result=error message if command=[object Object] and version=2 because command is an object
    ✓ should callback with success=false and result=error message if command=null and version=2 because command is null
    ✓ should callback with success=false and result=error message if command=true and version=2 because command is a boolean
    ✓ should callback with success=false and result=error message if command=false and version=2 because command is a boolean
    ✓ should callback with success=false and result=error message if command=getTCData and version=2.1 because version is a floating point number that is greater than 2
    ✓ should callback with success=false and result=error message if command=getTCData and version=2.1 because version is a floating point that doesn't evaluate to 2
    ✓ should callback with success=false and result=error message if command=getTCData and version=1 because version is not supported
    ✓ should callback with success=false and result=error message if command=getTCData and version=true because version is a boolean
    ✓ should callback with success=false and result=error message if command=getTCData and version=false because version is a boolean
    ✓ should callback with success=false and result=error message if command=getTCData and version=[object Object] because version is an object
    ✓ should callback with success=true and result=something if command=getTCData and version=2 because is a string but still 2
    ✓ should callback with success=true and result=something if command=getTCData and version=2.0 because is a string but still 2
    ✓ should callback with success=true and result=something if command=getTCData and version=2 because version is a floating point number that evaluates to 2
    ✓ should callback with success=true and result=something if command=getTCData and version=null because null is valid
    ✓ should callback with success=true and result=something if command=getTCData and version=0 because 0 is valid
    ✓ should throw an error if with callback undefined because it's undefined
    ✓ should throw an error if with callback true because it's a boolean
    ✓ should throw an error if with callback false because it's a boolean
    ✓ should throw an error if with callback [object Object] because it's an object
    ✓ should throw an error if with callback foo because it's a string
    ✓ should throw an error if with callback 2 because it's a number
    ✓ should call a custom command through the page interface
    ✓ should call a custom command through the page interface with multiple parameters
    ✓ should call a built-in ping command through the page interface and decorate the response in the middleware
    ✓ should call a built-in getTCData command through the page interface and decorate the response in the middleware
    ✓ should call a built-in getInAppTCData command through the page interface and decorate the response in the middleware
    ✓ should call a built-in getVendorList command through the page interface and decorate the response in the middleware
    ✓ should call a built-in addEventListener command through the page interface and decorate the response in the middleware
    ✓ should set gdprApplies=true, displayStatus="disabled", eventStatus="tcloaded", and cmpStatus="loaded" on the first set of the tcString when uiVisible=false
    ✓ should set gdprApplies=true, displayStatus="visible", eventStatus="cmpuishown", and cmpStatus="loaded" on the first set of the tcString when uiVisible=true
    ✓ should set gdprApplies=false, displayStatus="disabled", eventStatus="tcloaded", and cmpStatus="loaded" when the tcString is set to null
    ✓ should set gdprApplies=true, displayStatus="hidden", eventStatus="useractioncomplete", and cmpStatus="loaded" when the tcString is set a second time (39ms)
    ✓ should set gdprApplies=true, displayStatus="hidden", eventStatus="useractioncomplete", and cmpStatus="loaded" when the tcString is set a third time (52ms)

  command->PingCommand
    ✓ should return a Ping object (and nothing else) when called

  command->TCFCommand
    ✓ should have these values

  command->AddEventListenerCommand
    ✓ should immediately return a TCData object when called and queue a call as well

  command->RemoveEventListenerCommand
    ✓ should remove a queued TCDataCallback
    ✓ should return result=null and success=false for param=undefined
    ✓ should return result=null and success=false for param=1
    ✓ should return result=null and success=false for param=[object Object]
    ✓ should return result=null and success=false for param=
    ✓ should return result=null and success=false for param=1,3,5
    ✓ should return result=null and success=false for param=null
    ✓ should return result=null and success=false for param=banana
    ✓ should return result=null and success=false for param=function () {
        chai_1.expect.fail('this was never registered');
    }

  command->GetVendorListCommand
    ✓ should return version of TCModel.vendorListVersion if TCModel does not have a gvl and no parameter is passed
    ✓ should return latest GVL if "LATEST" is passed (52ms)
    ✓ should return version of GVL if integer is passed
    ✓ should cache a version of the gvl
    ✓ should return result=null and success=false for param=null
    ✓ should return result=null and success=false for param=[object Object]
    ✓ should return result=null and success=false for param=banana
    ✓ should return result=null and success=false for param=banana,apple,cake
    ✓ should return result=null and success=false for param=true
    ✓ should return result=null and success=false for param=false
    ✓ should return result=null and success=false for param=30000
    ✓ should return result=null and success=false for param=0
    ✓ should return result=null and success=false for param=1.1
    ✓ should return result=null and success=false for param=1.7,1.34

  command->GetInAppTCDataCommand
    ✓ should return a InAppTCData object when called
    ✓ should return result=null and success=false for param=null
    ✓ should return result=null and success=false for param=[object Object]
    ✓ should return result=null and success=false for param=banana
    ✓ should return result=null and success=false for param=banana,apple,cake
    ✓ should return result=null and success=false for param=1
    ✓ should return result=null and success=false for param=true
    ✓ should return result=null and success=false for param=0
    ✓ should return result=null and success=false for param=1.1
    ✓ should return result=null and success=false for param=1.7,1.34

  command->GetTCDataCommand
    ✓ should return a TCData object when called
    ✓ should return a TCData object and should not get a GVL if a tcString is set instead of tcModel
    ✓ should return a TCData object with a subset of vendors if a list of ids are passed
    ✓ should return result=null and success=false for param=null
    ✓ should return result=null and success=false for param=[object Object]
    ✓ should return result=null and success=false for param=banana
    ✓ should return result=null and success=false for param=banana,apple,cake
    ✓ should return result=null and success=false for param=1
    ✓ should return result=null and success=false for param=true
    ✓ should return result=null and success=false for param=false
    ✓ should return result=null and success=false for param=0
    ✓ should return result=null and success=false for param=1.1
    ✓ should return result=null and success=false for param=1.7,1.34

  command->CommandMap
    ✓ has these commands

  CallResponder
    ✓ should call a custom command before tcModel exists
    ✓ should call a built-in command only after tcModel exists
    ✓ should call a built-in custom command only after tcModel exists

  TCModel
    ✓ should construct a TCModel with no arguments
    ✓ should construct a TCModel with a GVL argument
    ✓ should clone a TCModel
    cmpId
      ✓ should be able to set and get a valid cmpId
      ✓ should not allow cmpId to be 1
      ✓ should not allow cmpId to be -1
      ✓ should not allow cmpId to be 1.1
    cmpVersion
      ✓ should be able to set and get a valid cmpVersion
      ✓ should not allow cmpVersion to be -1
      ✓ should not allow cmpVersion to be -1
      ✓ should not allow cmpVersion to be 1.1
    consentScreen
      ✓ should be able to set and get a valid consentScreen
      ✓ should not allow consentScreen to be -1
      ✓ should not allow consentScreen to be -1
      ✓ should not allow consentScreen to be 1.1
    created
      ✓ should be able to set and get a valid created
    lastUpdated
      ✓ should be able to set and get a valid lastUpdated
    isServiceSpecific
      ✓ should be able to set and get a valid isServiceSpecific
    useNonStandardStacks
      ✓ should be able to set and get a valid useNonStandardStacks
    purposeConsents
      ✓ should create an instance of Vector as purposeConsents on init
    purposeLegitimateInterests
      ✓ should create an instance of Vector as purposeLegitimateInterests on init
    vendorConsents
      ✓ should create an instance of Vector as vendorConsents on init
    vendorLegitimateInterests
      ✓ should create an instance of Vector as vendorLegitimateInterests on init
    specialFeatureOptins
      ✓ should create an instance of Vector as specialFeatureOptins on init
    consentLanguage
      ✓ should default to "EN" and set and get whatever I want
    setAllVendorConsents()
      ✓ should set all values
    unsetAllVendorConsents()
      ✓ should unset all values
    setAllVendorLegitimateInterests()
      ✓ should set all values
    unsetAllVendorLegitimateInterests()
      ✓ should unset all values
    setAllPurposeConsents()
      ✓ should set all values
    unsetAllPurposeConsents()
      ✓ should unset all values
    setAllPurposeLegitimateInterests()
      ✓ should set all values
    unsetAllPurposeLegitimateInterests()
      ✓ should unset all values
    setAllSpecialFeatureOptins()
      ✓ should set all values
    unsetAllSpecialFeatureOptins()
      ✓ should unset all values

  encoder/field->VectorEncodingType
    ✓ should have these values

  encoder/field->PurposeRestrictionVectorEncoder
    ✓ should return 12 0's for an empty PurposeRestrictionVector
    ✓ should create one restriction group and two vendor ranges
    ✓ should encode and decode and have the same output

  encoder/field->IntEncoder
    encode
      ✓ should encode an int and pad zeros to fill width
      ✓ should throw an error if the int is too large for the number of bits
      ✓ should throw an error if a negative int is passed
    decode
      ✓ should decode an int and pad zeros to fill width

  encoder/field->FieldEncoderMap
    ✓ Needs Unit Tests

  encoder/field->LangEncoder
    encode
      ✓ should encode "aa" into 12 as "000000000000"
      ✓ should encode "AA" into 12 as "000000000000"
      ✓ should encode "ba" into 12 as "000001000000"
      ✓ should encode "Ba" into 12 as "000001000000"
      ✓ should encode "zz" into 12 as "011001011001"
      ✓ should encode "ZZ" into 12 as "011001011001"
      ✓ should not encode "{Z" into 12: "{" is just after "Z"
      ✓ should not encode "-Z" into 12: "-" is before "A"
      ✓ should not encode "US" into 11: cannot encode into odd number of bits
      ✓ should not encode "ZZ" into 8: will not fit into 4 bits per letter
    decode
      ✓ should decode an encoded language
      ✓ should throw an error if the bit length is odd

  encoder/field->BooleanEncoder
    encode
      ✓ should encode true to 1
      ✓ should encode false to 0
    decode
      ✓ should decode 1 to true
      ✓ should decode 0 to false

  encoder/field->DateEncoder
    encode
      ✓ should encode a Date into 36 bits
      ✓ should not encode a Date into 26 bits
    decode
      ✓ should decode a Date

  encoder/field->VendorVectorEncoder
    encode
      ✓ should encode a vector of length 10 with no gaps as a bitfield
      ✓ should encode a vector of length 45 with no gaps as a bitfield
      ✓ should encode a vector with 46 vendors and no gaps as a range
      ✓ should encode a vector of length 48 1 gap as a bitfield since a that would be shorter
      ✓ should encode a vector of length 100 1 gap as a range since a that would be shorter
    decode
      ✓ should properly decode encoded vendor vector with 10 vendors and no gaps
      ✓ should properly decode encoded vendor vector with 100 vendors and no gaps
      ✓ should properly decode encoded vendor vector with 46 vendors and no gaps
      ✓ should properly decode encoded vendor vector with 10 vendors and 3 gaps
      ✓ should properly decode encoded vendor vector with 22 vendors and 6 gaps
      ✓ should properly decode encoded vendor vector with 50 vendors and 13 gaps
      ✓ should properly decode encoded vendor vector with 400 vendors and 100 gaps
      ✓ should decode a range with a single id

  encoder/field->FixedVectorEncoder
    encode
      ✓ should encode a vector
    decode
      ✓ should decode a fixed vector

  encoder->SegmentEncoder
    ✓ should encode a core segment
    ✓ TCModel->Core TC String->TCModel and should be equal
    ✓ should encode a OOB Vendors Allowed Segment properly into a string
    ✓ TCModel->OOBAllowed segment->TCModel and should be equal
    ✓ TCModel->PublisherTC segment->TCModel and should be equal with no custom purposes (never set)
    ✓ TCModel->PublisherTC segment->TCModel and should be equal with no custom purposes (empty object)
    ✓ TCModel->PublisherTC segment->TCModel and should be equal with 1 custom purpose
    ✓ TCModel->PublisherTC segment->TCModel and should be equal with 2 custom purpose
    ✓ TCModel->PublisherTC segment->TCModel and should be equal with 3 custom purpose, but with id gaps

  encoder/sequence->SegmentSequence
    ✓ v1:  isServiceSpecific=false, isForSaving=false, hasVendorsAllowed=false, and supportOOB=false
    ✓ v2:  isServiceSpecific=false, isForSaving=false, hasVendorsAllowed=false, and supportOOB=false
    ✓ v1:  isServiceSpecific=true, isForSaving=false, hasVendorsAllowed=false, and supportOOB=false
    ✓ v2:  isServiceSpecific=true, isForSaving=false, hasVendorsAllowed=false, and supportOOB=false
    ✓ v1:  isServiceSpecific=false, isForSaving=true, hasVendorsAllowed=false, and supportOOB=false
    ✓ v2:  isServiceSpecific=false, isForSaving=true, hasVendorsAllowed=false, and supportOOB=false
    ✓ v1:  isServiceSpecific=true, isForSaving=true, hasVendorsAllowed=false, and supportOOB=false
    ✓ v2:  isServiceSpecific=true, isForSaving=true, hasVendorsAllowed=false, and supportOOB=false
    ✓ v1:  isServiceSpecific=false, isForSaving=false, hasVendorsAllowed=true, and supportOOB=false
    ✓ v2:  isServiceSpecific=false, isForSaving=false, hasVendorsAllowed=true, and supportOOB=false
    ✓ v1:  isServiceSpecific=true, isForSaving=false, hasVendorsAllowed=true, and supportOOB=false
    ✓ v2:  isServiceSpecific=true, isForSaving=false, hasVendorsAllowed=true, and supportOOB=false
    ✓ v1:  isServiceSpecific=false, isForSaving=true, hasVendorsAllowed=true, and supportOOB=false
    ✓ v2:  isServiceSpecific=false, isForSaving=true, hasVendorsAllowed=true, and supportOOB=false
    ✓ v1:  isServiceSpecific=true, isForSaving=true, hasVendorsAllowed=true, and supportOOB=false
    ✓ v2:  isServiceSpecific=true, isForSaving=true, hasVendorsAllowed=true, and supportOOB=false
    ✓ v1:  isServiceSpecific=false, isForSaving=false, hasVendorsAllowed=false, and supportOOB=true
    ✓ v2:  isServiceSpecific=false, isForSaving=false, hasVendorsAllowed=false, and supportOOB=true
    ✓ v1:  isServiceSpecific=true, isForSaving=false, hasVendorsAllowed=false, and supportOOB=true
    ✓ v2:  isServiceSpecific=true, isForSaving=false, hasVendorsAllowed=false, and supportOOB=true
    ✓ v1:  isServiceSpecific=false, isForSaving=true, hasVendorsAllowed=false, and supportOOB=true
    ✓ v2:  isServiceSpecific=false, isForSaving=true, hasVendorsAllowed=false, and supportOOB=true
    ✓ v1:  isServiceSpecific=true, isForSaving=true, hasVendorsAllowed=false, and supportOOB=true
    ✓ v2:  isServiceSpecific=true, isForSaving=true, hasVendorsAllowed=false, and supportOOB=true
    ✓ v1:  isServiceSpecific=false, isForSaving=false, hasVendorsAllowed=true, and supportOOB=true
    ✓ v2:  isServiceSpecific=false, isForSaving=false, hasVendorsAllowed=true, and supportOOB=true
    ✓ v1:  isServiceSpecific=true, isForSaving=false, hasVendorsAllowed=true, and supportOOB=true
    ✓ v2:  isServiceSpecific=true, isForSaving=false, hasVendorsAllowed=true, and supportOOB=true
    ✓ v1:  isServiceSpecific=false, isForSaving=true, hasVendorsAllowed=true, and supportOOB=true
    ✓ v2:  isServiceSpecific=false, isForSaving=true, hasVendorsAllowed=true, and supportOOB=true
    ✓ v1:  isServiceSpecific=true, isForSaving=true, hasVendorsAllowed=true, and supportOOB=true
    ✓ v2:  isServiceSpecific=true, isForSaving=true, hasVendorsAllowed=true, and supportOOB=true

  encoder/sequence->FieldSequence
    ✓ should have this for v1
    ✓ should have this for v2

  Cloneable
    ✓ Needs Unit Tests

  GVL
    ✓ should fail to intialize without setting baseUrl
    ✓ should fail to set baseUrl to http://vendorlist.consensu.org/
    ✓ should fail to set baseUrl to https://vendorlist.consensu.org/ (secure url)
    ✓ should add a trailing slash to baseUrl if one is not there
    ✓ should propogate all values with passed in json
    ✓ should clone all values
    ✓ should get latest GVL if nothing is passed to the constructor
    ✓ should get latest GVL if "LATEST" is passed to the constructor
    ✓ should get versioned GVL if version number is passed
    ✓ should get versioned GVL if version number as string is passed
    ✓ should not re-request the "LATEST" vendorlist json if it has already downloaded it
    ✓ should not re-request a versioned vendorlist json if it has already downloaded it
    ✓ should not re-request a vendorlist json if it has been passed in to the constructor
    ✓ should not re-request a vendorlist json latest is request and then that version is requested later
    ✓ should narrow a group of vendors when narrowVendorsTo is called with list of ids
    ✓ should remove a vendor if it has a deletedDate
    ✓ should replace the language when changeLanguage() is called with valid language
    ✓ should throw an error if {Z is passed to changeLanguage()
    ✓ should throw an error if -Z is passed to changeLanguage()
    ✓ should throw an error if 35 is passed to changeLanguage()
    ✓ should throw an error if ZZZ is passed to changeLanguage()
    ✓ should throw an error if US is passed to changeLanguage()
    ✓ should throw an error if usa is passed to changeLanguage()
    ✓ should throw an error if .. is passed to changeLanguage()
    ✓ should throw an error if  EN is passed to changeLanguage()
    ✓ should throw an error if    is passed to changeLanguage()
    ✓ should throw an error if aa is passed to changeLanguage()
    ✓ should throw an error if aaa is passed to changeLanguage()
    ✓ should throw an error if zz is passed to changeLanguage()
    ✓ should throw an error if AA is passed to changeLanguage()
    ✓ should throw an error if ZZ is passed to changeLanguage()
    ✓ should throw an error if a is passed to changeLanguage()
    ✓ should throw an error if @# is passed to changeLanguage()
    ✓ should throw an error if 15 is passed to changeLanguage()
    ✓ should throw an error if {{ is passed to changeLanguage()
    ✓ should not request a file if the language is "EN"
    ✓ should not re-request the same language file if it has already been loaded
    ✓ should error if a 404 for the language file occurs
    ✓ should delete a language from the cache when requested
    ✓ should delete a gvl version from the cache when requested
    ✓ should group vendors by purpose consent
    ✓ should group vendors by purpose legInt
    ✓ should group vendors by purpose flexible
    ✓ should group vendors by special purpose 
    ✓ should group vendors by feature 
    ✓ should group vendors by special feature 

  TCString
    ✓ should throw an encoding error if no GVL is set
    ✓ should unset purposeLegitimateInterests 1 if it is set
    ✓ should set all vendorsDisclosed in the GVL when isServiceSpecific is false
    ✓ should produce only a publisherTC segment if the encoding option is passed
    ✓ should add a segment to a model on decode if a model is passed in (41ms)

  Json->fetch
    ✓ should fetch and parse a json with only a url
    ✓ should fail if a 404 returned
    ✓ should fail if a network error occurs

  Json->post
    ✓ should post and parse a json with a url and body
    ✓ should fail if a 404 returned
    ✓ should fail if a network error occurs

  Issues Reported
    ✓ 112 Vendor ranges incorrectly decoded from publisher restrictions
    ✓ 91 TCString.encode use 0 as vendorListVersion instead of gvl
    ✓ 77 should not delete tcfPolicyVersion and gvlSpecificationVersion after a language is changed
    ✓ 122 consentLanguage should not be ignored
    ✓ 117 TCString.encode writes all vendors as disclosed even after GVL.narrowVendorsTo([...])
    ✓ 142 it should throw an error if the bitfield length does not match the maxId
    ✓ 156 it should auto detect 1 version
    ✓ 162 Legal basis purpose restriction is reflected on vendors without flexible purposes too
    ✓ 204 Not possible to set vendorConsent for vendor which declared felxiblePurposes but no purposes
    ✓ 191 vendorConsents empty when decoding TCF 1.1 consentString that uses vendorRangeList
    ✓ 201 unable to decode valid TCF2 String

  model->Fields
    ✓ should be this

  model->ConsentLanguages
    ✓ should have only these languages


  325 passing (2s)
```

**Comment by chrispaterson** [comment]
I have a pre-merge hook that won't let me merge to master if the tests and build fail.

**Comment by mindplay-dk** [comment]
> I have a pre-merge hook that won't let me merge to master if the tests and build fail.

Locally? So it succeeds on your system - fails on mine, and fails with your CI script on the Travis server.

"Needs more" - do you need something else from me? 🙂

**Comment by mindplay-dk** [comment]
These are TS errors - are you using a global install of TypeScript by any chance?

Not having these deps locked-down seems a likely issue - type-checks may well have changed between versions.

**Comment by chrispaterson** [comment]
> > I have a pre-merge hook that won't let me merge to master if the tests and build fail.
> 
> Locally? So it succeeds on your system - fails on mine, and fails with your CI script on the Travis server.
> 
> "Needs more" - do you need something else from me? 🙂

I just put the "Needs more" label on it so I know that I'm waiting for a response, or I need to dig deeper to figure out what the problem is.



> These are TS errors - are you using a global install of TypeScript by any chance?
> 
> Not having these deps locked-down seems a likely issue - type-checks may well have changed between versions.

The repo should use the local typescript.  I don't have typescript installed globally.  When you run `yarn tsc --version` in the root does it give you 3.7.5?

**Comment by mindplay-dk** [comment]
> I just put the "Needs more" label on it so I know that I'm waiting for a response, or I need to dig deeper to figure out what the problem is.

Ah 🙂👍

> The repo should use the local typescript. I don't have typescript installed globally. When you run `yarn tsc --version` in the root does it give you 3.7.5?

yep

**Comment by mindplay-dk** [comment]
Do you for sure have the same (locked) deps installed?

That really seems like the only other explanation.

Could you try a clean check-out in a new folder and see if it still builds from there?

**Comment by chrispaterson** [comment]
As I said in [my previous comment](#issuecomment-689702754) I did a clean checkout of the repository.

**Comment by mindplay-dk** [comment]
Well, I did that, and the CI server presumably does that, so why does it build on some systems only? Remains a mystery. 😐

**Comment by chrispaterson** [comment]
I'm not sure, perhaps you have a global dependency that is interfering?  I'm going to close the issue because I don't think there's much I can do

**Comment by mindplay-dk** [comment]
Even if it's a global dependency problem, this project has an issue: it depends on an unchecked, global dependency - maybe it has an incorrect or missing version constraint?

If I run `npm list`, I'm seeing a long list of errors about `extraneous` packages - as mentioned, the only dependencies you have properly listed in `package.json` is `json`, `mocha` and `mocha-junit-reporter`. You seem to rely exclusively on `yarn` and the lock-file for versioning - I don't know much about yarn, but the lock-file is not an alternative to version constraints.

I would suggest you `npm list` and then `npm install {package} --save-dev` for any undeclared dependencies - I suspect that would solve the problem, but it should make this package more reliable overall, enables reliable updates of dependencies, and so on.

If there is a conflict with a globally installed dependency, proper version constraints should fix that - as the `package.json` looks right now, nothing would cause `npm` to check global dependencies at install time, and I'm betting `yarn` works the same.
